File: nQueens.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Program
{
    class Program
    {
        static void Main(string[] args)
        {
            Solution solution = new Solution();
            List<string> result = solution.solveQueens(4);
            foreach (string row in result)
            {
                Console.WriteLine(row);
            }
            Console.ReadKey();
        }
    }

    class Solution
    {
        HashSet<int> column = new HashSet<int>();
        HashSet<int> positiveDiag = new HashSet<int>();
        HashSet<int> negativeDiag = new HashSet<int>();
        List<string> result = new List<string>();

        public List<string> solveQueens(int n) 
        {
            string[,] board = new string[n, n];
            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    board[i, j] = ".";
                }
            }
            backtrack(0, n,board);
            return result;
        }

        private void backtrack(int r, int n, string[,] board)
        {
            if (r == n)
            {
                for (int i = 0; i < n; i++)
                {
                    StringBuilder sb = new StringBuilder();
                    for (int j = 0; j < n; j++)
                    {
                        sb.Append(board[i, j]);
                    }
                    result.Add(sb.ToString());
                }
                return;
            }

            for (int c = 0; c < n; c++)
            {
                if (column.Contains(c) || positiveDiag.Contains(r + c) || negativeDiag.Contains(r - c))
                    continue;
                column.Add(c);
                positiveDiag.Add(r + c);
                negativeDiag.Add(r - c);
                board[r, c] = "Q";
                backtrack(r + 1, n, board);
                // Clean up
                column.Remove(c);
                positiveDiag.Remove(r + c);
                negativeDiag.Remove(r - c);
                board[r, c] = ".";
            }
        }
    }
}

File: nQueens.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class nQueens {
  public static void main(String[] args) {
    Solution solution = new Solution();
    List<String> result = solution.solveQueens(4);
    for (String row : result) {
      System.out.println(row);
    }
  }
}

class Solution {
  Set<Integer> column = new HashSet<Integer>();
  Set<Integer> positiveDiag = new HashSet<Integer>();
  Set<Integer> negativeDiag = new HashSet<Integer>();
  List<String> result = new ArrayList<>();
  String[][] board;
  String[] copy;

  public List<String> solveQueens(int n) {
    board = new String[n][n];
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        board[i][j] = ".";
      }
    }
    backtrack(0, n);
    return result;
  }

  private void backtrack(int r, int n) {
    if (r == n) {
      copy = new String[board.length];
      for (String[] row : board) {
        String copy = String.join("", row);
        result.add(copy);
      }
      return;
    }

    for (int c = 0; c < n; c++) {
      if (column.contains(c) || positiveDiag.contains(r + c) || negativeDiag.contains(r - c))
        continue;
      column.add(c);
      positiveDiag.add(r + c);
      negativeDiag.add(r - c);
      board[r][c] = "Q";
      backtrack(r + 1, n);
      // Clean up
      column.remove(c);
      positiveDiag.remove(r + c);
      negativeDiag.remove(r - c);
      board[r][c] = ".";
    }
  }
}

/*
 * A better way to print
 * public static void main(String[] args) {
 * Solution solution = new Solution();
 * List<String> result = solution.solveQueens(4);
 * int count = 0;
 * for (String row : result) {
 * System.out.println(row);
 * count++;
 * if (count % 4 == 0) {
 * System.out.println();
 * }}}
 */
